ksqldb-functional-tests/src/main/java/io/confluent/ksql/test/tools/TestExecutor.java

```diff
@@ -499,53 +500,52 @@ private void processSingleRecord(
       final Set<Topic> possibleSinkTopics
   ) {
     final String topicName = producedRecord.topic();
-
-    final ConsumerRecord<byte[], byte[]> consumerRecord =
-        new org.apache.kafka.streams.test.ConsumerRecordFactory<>(
-            new ByteArraySerializer(),
-            new ByteArraySerializer()
-        ).create(
-            topicName,
-            producedRecord.key(),
-            producedRecord.value(),
-            producedRecord.timestamp()
-        );
-
-    testDriver.getTopologyTestDriver().pipeInput(consumerRecord);
+    final TestInputTopic<byte[], byte[]> inputTopic = testDriver
+        .getTopologyTestDriver()
+        .createInputTopic(topicName, new ByteArraySerializer(), new ByteArraySerializer());
+
+    inputTopic.pipeInput(
+        producedRecord.key(),
+        producedRecord.value(),
+        producedRecord.timestamp()
+    );
 
     final Topic sinkTopic = testDriver.getSinkTopic();
 
     processRecordsForTopic(
         testDriver.getTopologyTestDriver(),
-        sinkTopic
+        sinkTopic.getName()
     );
 
     for (final Topic possibleSinkTopic : possibleSinkTopics) {
       if (possibleSinkTopic.getName().equals(sinkTopic.getName())) {
         continue;
       }
+
       processRecordsForTopic(
           testDriver.getTopologyTestDriver(),
-          possibleSinkTopic
+          possibleSinkTopic.getName()
       );
     }
   }
 
   private void processRecordsForTopic(
-      final TopologyTestDriver topologyTestDriver,
-      final Topic sinkTopic
+      final TopologyTestDriver testDriver,
+      final String topicName
   ) {
-    while (true) {
-      final ProducerRecord<byte[], byte[]> producerRecord = topologyTestDriver.readOutput(
-          sinkTopic.getName(),
-          new ByteArrayDeserializer(),
-          new ByteArrayDeserializer()
-      );
+    final TestOutputTopic<byte[], byte[]> outputTopic = testDriver
+          .createOutputTopic(topicName, new ByteArrayDeserializer(), new ByteArrayDeserializer());
 
-      if (producerRecord == null) {
+    while (true) {
+      if (outputTopic.isEmpty()) {
         break;
       }
-
+      final TestRecord<byte[], byte[]> testRecord = outputTopic.readRecord();
+      final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord(
+          topicName,
+          testRecord.getKey(),
+          testRecord.getValue()
```

Comment:
We also need to pass `testRecord.timestamp()` into the `ProducerRecord` to be able to verify result timestamps.